presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -18,61 +18,37 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.ImmutableSet;
 
+import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
-import java.util.Set;
 
 import static com.facebook.presto.cassandra.util.CassandraCqlUtils.toCQLCompatibleString;
-import static com.google.common.collect.Sets.cartesianProduct;
 import static java.util.Objects.requireNonNull;
 
 public class CassandraClusteringPredicatesExtractor
 {
-    private final List<CassandraColumnHandle> clusteringColumns;
-    private final TupleDomain<ColumnHandle> predicates;
     private final ClusteringPushDownResult clusteringPushDownResult;
+    private final TupleDomain<ColumnHandle> predicates;
 
-    public CassandraClusteringPredicatesExtractor(List<CassandraColumnHandle> clusteringColumns, TupleDomain<ColumnHandle> predicates)
+    public CassandraClusteringPredicatesExtractor(List<CassandraColumnHandle> clusteringColumns, TupleDomain<ColumnHandle> predicates, String cassandraVersion)
     {
-        this.clusteringColumns = ImmutableList.copyOf(requireNonNull(clusteringColumns, "clusteringColumns is null"));
         this.predicates = requireNonNull(predicates, "predicates is null");
-        this.clusteringPushDownResult = getClusteringKeysSet(clusteringColumns, predicates);
+        this.clusteringPushDownResult = getClusteringKeysSet(clusteringColumns, predicates, cassandraVersion);
     }
 
     public List<String> getClusteringKeyPredicates()
     {
-        Set<List<Object>> pushedDownDomainValues = clusteringPushDownResult.getDomainValues();
-
-        if (pushedDownDomainValues.isEmpty()) {
+        if (clusteringPushDownResult.getDomainQuery().isEmpty()) {
             return ImmutableList.of();
         }
-
-        ImmutableList.Builder<String> clusteringPredicates = ImmutableList.builder();
-        for (List<Object> clusteringKeys : pushedDownDomainValues) {
-            if (clusteringKeys.isEmpty()) {
-                continue;
-            }
-
-            StringBuilder stringBuilder = new StringBuilder();
-
-            for (int i = 0; i < clusteringKeys.size(); i++) {
-                if (i > 0) {
-                    stringBuilder.append(" AND ");
-                }
-
-                stringBuilder.append(CassandraCqlUtils.validColumnName(clusteringColumns.get(i).getName()));
-                stringBuilder.append(" = ");
-                stringBuilder.append(CassandraCqlUtils.cqlValue(toCQLCompatibleString(clusteringKeys.get(i)), clusteringColumns.get(i).getCassandraType()));
-            }
-
-            clusteringPredicates.add(stringBuilder.toString());
-        }
-        return clusteringPredicates.build();
+        ImmutableList.Builder<String> clusteringColumnSql = ImmutableList.builder();
+        return clusteringColumnSql.add(Joiner.on(" AND ").join(clusteringPushDownResult.getDomainQuery())).build();
```

Comment:
Changes completed. Now `getClusteringKeyPredicates` also will return single `String` instead of `List<String>`.